OpenKeychain/src/main/java/org/sufficientlysecure/keychain/securitytoken/SecurityTokenConnection.java

```diff
@@ -304,6 +309,140 @@ public void clearSecureMessaging() {
 
     // region pin management
 
+    private byte[] executeKdf(byte[] pin, int password_type) throws IOException {
```

Comment:
This is a long function, that does two separate things:
* retrieve and parse the KDF DO
* calculate actual kdf based on parameters

I suggest splitting those parts up, as such:
```
byte[] calculateKdfIfNecessary(byte[] pin, int type) {
  if (!capabilities.isHasKdf) return
  KdfParameters kdfParameters = retrieveKdfDo(); // can cache internally
  return KdfCalculator.calculateKdf(kdfParameters.forType(type), pin);
}
```

the actual kdf calculation is a pure function, and we should definitely test it. there is a test vector in the [opgnpgp card spec, page 20](https://gnupg.org/ftp/specs/OpenPGP-smart-card-application-3.4.pdf).

Reply:
I've split the function up into smaller, more sensible units & added a test for the KDF calculation.
Any comment on the resulting code would be greatly appreciated :slightly_smiling_face: